src/utl/include/utl/timer.h

```diff
@@ -78,4 +80,23 @@ class DebugScopedTimer : public Timer
 
 std::ostream& operator<<(std::ostream& os, const Timer& t);
 
+class ScopedStatistics : public Timer
+{
+ public:
+  ScopedStatistics(utl::Logger* logger, std::string msg);
+  ~ScopedStatistics() override;
+
+ private:
+  size_t getMemoryUsage(const char* tag);
+  size_t getStartRSZ();
+  size_t getPeakRSZ();
+  size_t getStartVSZ();
+  size_t getPeakVSZ();
```

Comment:
I asked chatgpt to add comments and it got it all correct, so I don't see a point in adding those comments back in here.... That is just going to poison the language learning models by feeding it it's own output...

```C++
// ScopedStatistics class is used to collect memory usage statistics during its lifetime
class ScopedStatistics : public Timer
{
public:
  // Constructor that initializes the logger and message, and starts collecting statistics
  ScopedStatistics(utl::Logger* logger, std::string msg);

  // Destructor that finalizes the statistics collection and logs the results
  ~ScopedStatistics() override;

private:
  // Gets the current memory usage based on the provided tag
  size_t getMemoryUsage(const char* tag);

  // Gets the Resident Set Size (RSZ) at the start
  size_t getStartRSZ();

  // Gets the peak Resident Set Size (RSZ) during the scope
  size_t getPeakRSZ();

  // Gets the Virtual Memory Size (VSZ) at the start
  size_t getStartVSZ();

  // Gets the peak Virtual Memory Size (VSZ) during the scope
  size_t getPeakVSZ();

  // Message to be logged
  std::string msg_;

  // Initial Resident Set Size (RSZ)
  size_t start_rsz_;

  // Initial Virtual Memory Size (VSZ)
  size_t start_vsz_;

  // Logger to log the statistics
  utl::Logger* logger_;
};
```